Add resources for aws_billing_report and aws_billing_reports.

miah opened the pull request:
This PR adds two resources, `aws_billing_report` and `aws_billing_reports`, tests, and documentation.

With `aws_billing_report` you can search for specific reports and test its properties.

```
describe aws_billing_report('inspec1') do
  its('report_name') { should cmp 'inspec1' }
  its('time_unit') { should cmp 'DAILY' }
end
```

`aws_billing_reports` uses a FilterTable to store all reports and their properties. 
```
describe aws_billing_reports do
  its('report_name') { should include 'inspec1' }
  its('s3_bucket') { should include 'inspec1-s3-bucket' }
end
```

You can also use `.where` to match multiple entries.
```
describe aws_billing_reports.where { report_name =~ /inspec.*/ } do
  its('time_unit') { should_not include 'DAILY' }
  its('s3_bucket') { should include 'inspec1-s3-bucket' }
end
```

I was not able to add integration testing through Terraform as it lacks the resources to create billing reports.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

Fixes https://github.com/chef/inspec/issues/2720

miah committed:
Add resources for aws_billing_report and aws_billing_reports.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+## Propery Examples
+
+### report_name
+ The reports name.
```
Typo: missing possessive apostrophe

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+  end
+
+### time_unit
+ The interval of time covered by the report. Valid values: HOURLY, or Daily.
```
Is the upper/titlecase difference present in the API?  I feel like we should protect the user from that.

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+ The interval of time covered by the report. Valid values: HOURLY, or Daily.
+
+  describe aws_billing_report('inspec1') do
+    its('time_unit') { should cmp 'HOURLY' }
```
In fact, as a bit of sugar, we could add matchers `be_hourly` and `be_daily`.

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+ The prefix that AWS adds to the report when stored.
+
+  describe aws_billing_report('inspec1') do
+    its('s3_prefix') { should cmp 'inspec1' }
```
Might add some slashes to your example, so it looks more like a typical S3 prefix.

`its('s3_prefix') { should cmp 'inspec1/accounting/' }`

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+ The list of manifests created for the report. Valid values: REDSHIFT, or QUICKSIGHT.
+
+  describe aws_billing_report('inspec1') do
+    its('additional_artifacts') { should cmp 'QUICKSIGHT' }
```
Is this an array?  If so, we should probably use include (yes, `cmp` will unpack a single-element array, but if it is 2-element, it would fail). 

`its('additional_artifacts') { should include 'QUICKSIGHT' }`

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+  end
+
+## Matchers
+
```
If you decide not to add the be_daily and be_hourly matchers, it would be great to add a note that this resource has no resource-specific matchers, but you can use the universal matchers...

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+# aws\_billing\_reports
+
+Use the `aws_billing_reports` InSpec audit resource to test properties of a some or all AWS Cost and Billing reports.
+
```
Might mention using the singular form to audit a billing report in detail.

clintoncwolfe commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,98 @@
+# aws\_billing\_report
+
+Use the `aws_billing_report` InSpec audit resource to test properties of a single AWS Cost and Billing report.
+
```
Might mention the plural resource for auditing in bulk.

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+
+  # Use the .where clause to match a property to one or more rules in the available reports.
+  describe aws_billing_reports.where { report_name =~ /inspec.*/ } do
+    its('report_name') { should include 'inspec1' }
```
Properties on plural resources should generally be plural, unless it is a property of the entire group.

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+## Propery Examples
+
+### report_name
+ The reports name.
```
"A list containing the names of the reports matched by the resource."

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+### time_unit
+ The interval of time covered by the report. Valid values: HOURLY, or Daily.
```
"A list of intervals covered by the reports."

To me, this property seems difficult to use.  Some matchers based on the data might be handy, though:

it { should all_be_hourly } # This read well if you replace "it" with "they"
it { should_not have_any_be_daily }

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+### compression
+ The reports compression type. Valid values: ZIP, or GZIP.
```
As above, I think this would be more handy as a matcher.  `it { should all_have_zip_compression }` ?

clintoncwolfe commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+### s3_bucket
+ The s3_bucket the report is stored in.
```
A list of the S3 buckets the reports are stored in.

clintoncwolfe commented on lib/resources/aws/aws_billing_report.rb:
```diff
@@ -0,0 +1,79 @@
+require 'utils/filter'
```
Not that this matters much, but you only need to pull this in if you're using FilterTable.

clintoncwolfe commented on lib/resources/aws/aws_billing_report.rb:
```diff
@@ -0,0 +1,79 @@
+    )
+
+    if validated_params.empty?
+      raise ArgumentError, "You must provide the parameter 'report_definition' to aws_billing_report."
```
Is `report_definition` the same as `report_name`?

clintoncwolfe commented on lib/resources/aws/aws_billing_report.rb:
```diff
@@ -0,0 +1,79 @@
+
+  def report
+    definitions = backend.describe_report_definitions.report_definitions
+    report = definitions.select { |r| r.report_name.eql?(@report_definition) }
```
Fun fact: `Enumerable.detect` will return the first matching element (like `select { }.first`)

clintoncwolfe commented on lib/resources/aws/aws_billing_report.rb:
```diff
@@ -0,0 +1,79 @@
+  private
+
+  def report
+    definitions = backend.describe_report_definitions.report_definitions
```
We need to add a pagination loop here.  [Interface](https://docs.aws.amazon.com/sdkforruby/api/Aws/CostandUsageReportService/Client.html#describe_report_definitions-instance_method) , [example](https://github.com/chef/inspec/blob/master/lib/resources/aws/aws_iam_policies.rb#L33)

clintoncwolfe commented on lib/resources/aws/aws_billing_reports.rb:
```diff
@@ -0,0 +1,65 @@
+  desc 'Verifies settings for AWS Cost and Billing Reports.'
+  example "
+      describe aws_billing_reports do
+        its('report_name') { should include 'inspec1' }
```
If you pluralize the property names, reminder to update in-code example.

clintoncwolfe commented on lib/resources/aws/aws_billing_reports.rb:
```diff
@@ -0,0 +1,65 @@
+  def fetch_from_api
+    @table = []
+    backend = BackendFactory.create(inspec_runner)
+    backend.describe_report_definitions.report_definitions.each do |r|
```
We need to add a pagination loop here.  [Interface](https://docs.aws.amazon.com/sdkforruby/api/Aws/CostandUsageReportService/Client.html#describe_report_definitions-instance_method) , [example](https://github.com/chef/inspec/blob/master/lib/resources/aws/aws_iam_policies.rb#L33)

clintoncwolfe commented on test/unit/resources/aws_billing_report_test.rb:
```diff
@@ -0,0 +1,83 @@
+    r = AwsBillingReport.new('inspec1')
+    assert_equal('inspec1', r.report_name)
+    assert_equal('HOURLY', r.time_unit)
+    assert_equal('textORcsv', r.format)
```
Looks like format is a method, but is undocumented.

clintoncwolfe commented on test/unit/resources/aws_billing_report_test.rb:
```diff
@@ -0,0 +1,83 @@
+    assert_equal('inspec1', r.s3_prefix)
+    assert_equal('us-east-1', r.s3_region)
+    assert_equal(['REDSHIFT'], r.additional_artifacts)
+    assert_equal(['RESOURCES'], r.additional_schema_elements)
```
This appears to be undocumented, as well.  Remember, if we're not sure of its semantics, we can leave it out and add it later.

clintoncwolfe reviewed (dismissed):
Great work, @miah!  I love seeing singular and plural resources arrive at the same time, and this was a tough nut to crack.  Some docs/examples tweaks; mainly around pluralizing the properties on the plural resource.  Also, we need to be sure to paginate the results from the API call.  Finally, even though we can't get TF to make reports, can we make reports ourselves (and document in TESTING_AGAINST_AWS.md), then have a integration test?  Again, overall, very solid - thanks so much for your contribution!

miah committed:
Support pagination in aws_billing_report and aws_billing_reports.

Add paginated testing backend. It wasn't possible to enable pagination in the
aws-sdk stub interface.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe committed:
Fix failing unit test by caching the backend

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Adjust resource param name to be consistent

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Update FilerTable API, and use plural property names on a plural resource

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Use deduplication on most fields on plural

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Make docs and code consistent on case of compression and timeunit

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

jerryaldrichiii commented on docs/resources/aws_billing_report.md.erb:
```diff
@@ -0,0 +1,106 @@
+    it { should be_hourly }
+  end
+
+### be_daily
```
Could we remove this or add an example.

jerryaldrichiii commented on docs/resources/aws_billing_reports.md.erb:
```diff
@@ -0,0 +1,93 @@
+
+### compressions
+
+A list of the compression types of the reports matched by the query.  Valid values: `zip`, or `gzip`. This list is de-duplicated, so its count may not match the query count.
```
Two spaces here before `Valid`, doesn't match the rest.

jerryaldrichiii commented on test/unit/resources/aws_billing_backend.rb:
```diff
@@ -0,0 +1,99 @@
+          report_name: 'inspec1',
+          time_unit: 'HOURLY',
+          format: 'textORcsv',
+          compression: 'ZIP',
```
Should this and below be lowercase `zip`/`gzip`?

clintoncwolfe replied:
All caps is correct per https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/describe-report-definitions.html

jerryaldrichiii reviewed (dismissed):
Great work! Almost got it merged! Only 2 small docs things and a question.

clintoncwolfe committed:
Doc edits per feedback

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe reviewed (approved):
I had previously reviewed and accepted this, with the caveat about a broekn unit test.  I fixed that; let's ship this.
